.github/workflows/check-whitespace.yml

```diff
@@ -13,38 +13,57 @@ jobs:
   check-whitespace:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
       with:
         fetch-depth: 0
 
     - name: git log --check
       id: check_out
       run: |
-        log=
+        problems=()
         commit=
-        while read dash etc
+        commitText=
+        lastcommit=
+        while read dash sha etc
         do
           case "${dash}" in
           "---")
-            commit="${etc}"
+            if test -z "${commit}"
+            then
+              lastcommit=${sha}
+            fi
+            commit="${sha}"
+            commitText="${sha} ${etc}"
```

Comment:
Oh, I also thought of something different: We could also consider using some markup in the summary, i.e. `commitText="[${sha}](https://github.com/${{ github.repository }}/commit/${sha}) ${etc}"`? It would probably make sense to do this _only_ for the summary, where GitHub-flavored Markdown is supported, and not in the job's log, where Markdown is not supported.

However, what _is_ supported in the job's log (but not in the summary): [workflow commands, e.g. to display errors](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message). We could even use the feature where the exact file and line are specified. Something like:

```sh
echo "::error file=${dash%:*},line=${dash##*:}::${sha} ${etc} (introduced in ${commit})"
```

@webstech what do you think about that idea?

Reply:
That `echo "::error [...]"` command should probably be guarded by something like

```sh
case "${dash}" in
*:[1-9]*) # contains file and line number information
  [... show the error as specified above ...]
  ;;
*)
  [... show "${dash} ${sha} ${etc}" verbatim
  ;;
esac
```

Reply:
> what do you think about that idea?

I will play with both of them to see how it looks.  May not need both.

> `*:[1-9]*)`

Perhaps `*:[0-9]+)`.

Reply:
> Perhaps `*:[0-9]+)`.

It's an fnmatch pattern, not a regex... So `[0-9]+` would only match a digit followed by a plus sign. 😁

Reply:
> It's an fnmatch pattern, not a regex

You are correct (of course).  bash made me painfully aware of that. 😒

I am not really happy with using `::error`.  In fact I am not happy at all.  The link in the email for a failing job takes you to the summary (yay) but the `details` link in the PR takes you to the actual failing step output.  The `::error` message in the step output is not really very useful.  The job summary is where the links are active. 

Using your suggestion (I had just read about using md in the summary when working on this), I have a fresh [sample](https://github.com/webstech/check-whitespace/actions/runs/3701398690) showing both annotations and summary.  The summary errors are linked to the code.  

My preference is to use the summary and not create annotations.  To overcome the `details` link issue, a message could be added to the step log to check the summary for fix information. @dscho What do you think?

Reply:
> The summary errors are linked to the code.

Hrm. If the file in question was modified in the PR, I _think_ the link would lead to the correct line. But here it looks as if it does not, as `trailing.txt` is not even modified in the changes where the annotation lives.

So I guess my feedback was not helpful at all, and you should go with what your guts told you all along, @webstech. Sorry about that!

Reply:
Thank you!